tests/equality/equality.cpp

```diff
@@ -0,0 +1,995 @@
+
+    EXPECT_TRUE(u1->equal(u2));
+}
+
+TEST(Equality, unitsNotEqualDifferingNumberOfUnits)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    libcellml::ImportSourcePtr is = libcellml::ImportSource::create();
+
+    u1->addUnit("second", 1.0, "second_id");
+    u1->addUnit("meter", -1.0, "meter_id");
+    u2->addUnit("second", 1.0, "second_id");
+
+    EXPECT_FALSE(u1->equal(u2));
+}
+
+TEST(Equality, unitsNotEqualDifferingUnitsBasic)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    libcellml::ImportSourcePtr is = libcellml::ImportSource::create();
+
+    u1->addUnit("second", 1.0, "unit_id");
+    u2->addUnit("second", 2.0, "unit_id");
+
+    EXPECT_FALSE(u1->equal(u2));
+}
+
+TEST(Equality, unitsEqualDifferingOrderOfUnits)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("second", 1.0, "unit_id");
+    u1->addUnit("meter", 3.0);
+    u2->addUnit("meter", 3.0);
+    u2->addUnit("second", 1.0, "unit_id");
+
+    EXPECT_TRUE(u1->equal(u2));
+}
+
+TEST(Equality, unitsEqualDifferingOrderOfUnitsManyUnit)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("second", 1.0, "unit_id");
+    u1->addUnit("meter", 3.0);
+    u1->addUnit("apple", 1.0);
+    u1->addUnit("jedi", "yota", -2.0);
+    u1->addUnit("flip");
+    u1->addUnit("lumen");
+
+    u2->addUnit("jedi", "yota", -2.0);
+    u2->addUnit("second", 1.0, "unit_id");
+    u2->addUnit("lumen");
+    u2->addUnit("meter", 3.0);
+    u2->addUnit("apple", 1.0);
+    u2->addUnit("flip");
+
+    EXPECT_TRUE(u1->equal(u2));
+}
+
+TEST(Equality, unitsNotEqualDifferingOrderOfUnits)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("meter", 3.0);
+    u1->addUnit("second", 1.0, "unit_id");
+    u2->addUnit("second", 2.0, "unit_id");
+    u2->addUnit("meter", 3.0);
+
+    EXPECT_FALSE(u1->equal(u2));
+}
+
+TEST(Equality, unitsEqualPrecision1)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("second", 1.0);
+    u2->addUnit("second", 1.0 + std::numeric_limits<double>::epsilon());
+
+    EXPECT_TRUE(u1->equal(u2));
+}
+
+TEST(Equality, unitsEqualPrecision2)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    u1->addUnit("second", 1.0);
+    u2->addUnit("second", 1.0 + 2 * std::numeric_limits<double>::epsilon());
+
+    EXPECT_FALSE(u1->equal(u2));
+}
+
+TEST(Equality, unitsEqualPrecision3)
+{
```

Comment:
If it fails with 2x, why do you need to test 3x as well?

Reply:
82ed235